src/models/losses.py

```diff
@@ -4,6 +4,327 @@
 import torch
 from torch import nn
 
+# pylint: disable=too-many-arguments
+# pylint: disable=too-many-locals
+
+
+def generator_loss(
+    logits: dict[str, dict[str, torch.Tensor]],
+    inception_encoder: Any,
+    fake_imgs: torch.Tensor,
+    real_labels: torch.Tensor,
+    words_emb: torch.Tensor,
+    sent_emb: torch.Tensor,
+    match_labels: torch.Tensor,
+    cap_lens: torch.Tensor,
+    class_ids: torch.Tensor,
+    vgg_encoder: Any,
+    real_imgs: torch.Tensor,
+    device: Any,
+    const_dict: dict[Any, Any],
+) -> Any:
+    """Calculate the loss for the generator.
+
+    Args:
+        logits: Dictionary with fake/real and word-level/uncond/cond logits
+
+        inception_encoder: The inception encoder model.
+
+        fake_imgs: The fake images generated by the generator. [shape: (batch_size, 3, 256, 256)]
+
+        real_labels: Label for "real" image as predicted by discriminator,
+        this is a tensor of ones. [shape: (batch_size, 1)].
+
+        words_emb: The embeddings for all the words in the captions.
+        shape: (batch_size, embedding_size, max_caption_length)
+
+        sent_emb: The embeddings for the sentences.
+        shape: (batch_size, embedding_size)
+
+        match_labels: Tensor of shape: (batch_size, 1).
+        This is of the form torch.tensor([0, 1, 2, ..., batch-1])
```

Comment:
true, but all the labels [real_labels, fake_labels, match_labels] are handled by a separate function called prepare_labels, and it does not make sense to generate labels inside this discriminator loss, instead we generate at one central place and pass it in the training loop to different places

Reply:
umm but why should the losses part handle the label creation logic? even real labels is just a label of ones and fake labels is label of 0's, but we do pass it in downstream functions... I mean you are right that function notation will have less argument and it can be created easily, but i'm confused like if you tomorrow want to change the logic of match labels you have to change it everywhere where it is used, currently there is just a single source of truth for labels. let me know your thoughts.

Reply:
> even real labels is just a label of ones and fake labels is label of 0's, but we do pass it in downstream functions

that makes sense. resolved.